const content = {
  pageType: "jh-page",
  pageId: "exampleClass",
  table: "example_class",
  pageName: "班级页面",
  // 1table 外需要添加的其他 resource
  resourceList: [
    // **** //
  ],
  includeList: [
    // **** //
  ], // { type: < js | css | html >, path: ''}
  common: {
    data: {},
    dataString: "<<!DATA!>>",
    dataExpression: {},
    doUiAction: {},
    watch: "<<!WATCH!>>",
    methods: "<<!METHODS!>>",
    computed: "<<!COMPUTED!>>",
    mountedString: "<<!MOUNTED!>>",
    createdString: "<<!CREATED!>>",
  },
  headContent: [],
  pageContent: [],
  actionContent: [],
  style: "",
};

module.exports = content;
